core/src/main/java/io/questdb/cutlass/http/WaitProcessor.java

```diff
@@ -0,0 +1,211 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cutlass.http;
+
+import io.questdb.cairo.pool.ex.EntryUnavailableException;
+import io.questdb.mp.*;
+import io.questdb.std.time.MillisecondClock;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.PriorityQueue;
+import java.util.concurrent.locks.LockSupport;
+
+public class WaitProcessor extends  SynchronizedJob implements RescheduleContext {
+
+    private static final int retryQueueLength = 4096;
+    private final RingQueue<RetryHolder> inQueue = new RingQueue<>(RetryHolder::new, retryQueueLength);
+    private final Sequence inPubSequence = new MPSequence(retryQueueLength);
+    private final Sequence inSubSequence = new SCSequence();
+    private final PriorityQueue<RetryHolder> nextRerun = new PriorityQueue<>(64);
+    private final RingQueue<RetryHolder> outQueue = new RingQueue<>(RetryHolder::new, retryQueueLength);
+    private final Sequence outPubSequence = new SPSequence(retryQueueLength);
+    private final Sequence outSubSequence = new MCSequence(retryQueueLength);
+    private final MillisecondClock clock;
+    private final long maxWaitCapMs;
+    private final double exponentialWaitMultiplier;
+
+    public WaitProcessor(WaitProcessorConfiguration configuration) {
+        this.clock = configuration.getClock();
+        this.maxWaitCapMs = configuration.getMaxWaitCapMs();
+        this.exponentialWaitMultiplier = configuration.getExponentialWaitMultiplier();
+
+        inPubSequence.then(inSubSequence).then(inPubSequence);
+        outPubSequence.then(outSubSequence).then(outPubSequence);
+    }
+
+    @Override
+    protected boolean runSerially() {
+        return processInQueue() || sendToOutQueue();
+    }
+
+    @Override
+    // This supposed to run in http execution thread / job
+    public void reschedule(Retry retry) {
+        reschedule(retry, 0, 0);
+    }
+
+    private void reschedule(Retry retry, int attempt, long waitStartMs) {
+        long now = clock.getTicks();
+        retry.getAttemptDetails().attempt = attempt;
+        retry.getAttemptDetails().lastRunTimestamp = now;
+        retry.getAttemptDetails().waitStartTimestamp = attempt == 0 ? now : waitStartMs;
+
+        while (true) {
+            long cursor = inPubSequence.next();
+            // -2 = there was a contest for queue index and this thread has lost
+            if (cursor < -1) {
+                continue;
+            }
+
+            // -1 = queue is empty. It means there are already too many retries waiting
+            // Send error to client.
+            if (cursor < 0) {
+                throw EntryUnavailableException.INSTANCE;
```

Comment:
Is this tested?

I could be wrong here, but when this exception is thrown out of `runReruns` it would be not be handled at all?